Fix #9635: Marks translation suggestions as rejected when an exploration is removed from a story.

kaylahardie opened the pull request:
1. This PR fixes #9635 .
2. This PR does the following: This PR is an extension of my [PR 10022](https://github.com/oppia/oppia/pull/10022). It marks translation suggestions as rejected when an exploration is removed from a story. Tests are added to make sure the translation suggestions are successfully rejected.

**Reviewers can start from commit e2cab52**

## Essential Checklist

- [ x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ x] The linter/Karma presubmit checks have passed locally on your machine.
- [ x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [ x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

kaylahardie committed:
marks translation suggestions rejected when a topic is deleted.

kaylahardie committed:
Merge remote-tracking branch 'upstream/develop' into mark-translations-rejected-for-topic-deletion

kaylahardie committed:
fixed whitespace.

kaylahardie committed:
Improved documentation.

kaylahardie committed:
Made testing code more modular for future tests.

kaylahardie committed:
Minor word change.

kaylahardie committed:
marks translation suggestions rejected if an exp is removed from a story

DubeySandeep commented on core/domain/story_services.py:
```diff
@@ -452,6 +453,10 @@ def update_story(
     create_story_summary(new_story.id)
     opportunity_services.update_exploration_opportunities(old_story, new_story)
+    for exp_id in exp_ids_removed_from_story:
+        suggestion_services.reject_translation_suggestions_with_exp_target_id(
```
This will make n*m db calls?

n = number of exps deleted in story
m = number of suggestions in each exploration.

If that's tru then try to make it in 1 db calls if possible? (Use put_multi)

kaylahardie replied:
Done (I use put_multi in my reject_suggestions method)

DubeySandeep commented on core/domain/story_services.py:
```diff
@@ -518,9 +523,13 @@ def delete_story(committer_id, story_id, force_deletion=False):
+    # exploration used in the story.
     opportunity_services.delete_exploration_opportunities(exp_ids)
+    for exp_id in exp_ids:
+        suggestion_services.reject_translation_suggestions_with_exp_target_id(
```
DItto as above*

kaylahardie replied:
Done!

DubeySandeep commented on core/domain/suggestion_services.py:
```diff
@@ -323,6 +323,30 @@ def reject_question_suggestions_with_skill_target_id(skill_id):
+        exp_id: The exploration ID corresponding to the target ID of the
+            translation suggestion.
+    """
+    suggestions = query_suggestions(
```
Do we have specific function to fetch translation suggestions if yes then use that (we had planned of removing this query function in future!)

kaylahardie replied:
I made a function for fetching translation suggestions with given exploration ids called get_translation_suggestions_with_exp_ids(exp_ids) in the gae_models.py and in suggestion_services.py

DubeySandeep commented on core/domain/suggestion_services.py:
```diff
@@ -323,6 +323,30 @@ def reject_question_suggestions_with_skill_target_id(skill_id):
+        ]
+    )
+    for suggestion in suggestions:
+        reject_suggestion(
```
Create a reject sugestion**s** and reject all suggestions at once!

kaylahardie replied:
Done

DubeySandeep commented on core/storage/suggestion/gae_models.py:
```diff
@@ -101,6 +101,12 @@
+# associated with an exploration that no longer corresponds to the story.
+# The story could be deleted or the exploration could have been
+# removed from the story.
+INVALID_STORY_REJECT_MESSAGE = 'The associated story no longer exists.'
```
hmm.. maybe exploration (not story) as if we remove a chapter from the story then also we reject the suggestions, right?

kaylahardie replied:
Can you clarify what you mean by this? When you say remove a chapter you mean removing the exploration from the story right?...Sometimes the suggestion is invalid because a topic or story is deleted and other times it is because the exploration is removed from the story. I changed the wording, does it help make it clearer?

DubeySandeep reviewed (changes requested):
@kaylahardie Took a quick pass and left a few comments, PTAL!

kaylahardie committed:
Fixed it to make less storage calls.

kaylahardie committed:
Merge remote-tracking branch 'upstream/develop' into mark-translations-rejected-for-topic-deletion

kaylahardie committed:
Fixed whitespace.

kaylahardie committed:
Fixed one more whitepsace

kaylahardie committed:
Merge branch 'mark-translations-rejected-for-topic-deletion' into mark-translation-rejected-when-exp-removed-from-story

kaylahardie committed:
Addressed reviewers comments.

kaylahardie committed:
Added check and test for the empty exp_id case.

kaylahardie committed:
Merge branch 'mark-translations-rejected-for-topic-deletion' into mark-translation-rejected-when-exp-removed-from-story

kaylahardie reviewed (commented):
@DubeySandeep thanks so much for your feedback :) It really helped me think more deeply about the storage aspects. I made some changes PTAL.

kaylahardie committed:
Merge remote-tracking branch 'upstream/develop' into mark-translations-rejected-for-topic-deletion

kaylahardie committed:
Improved the get_translation_suggestions_with_exp_ids test and responded to reviewers comments.

kaylahardie committed:
Fixed whitespace.

kaylahardie committed:
Changed the return of get_content_html for MockExploration because it had unused arguments.

kaylahardie committed:
Merge branch 'mark-translations-rejected-for-topic-deletion' into mark-translation-rejected-when-exp-removed-from-story

kaylahardie committed:
Added a comment to be more clear for get_content_html.

kaylahardie committed:
Merge remote-tracking branch 'upstream/develop' into mark-translations-rejected-for-topic-deletion

kaylahardie committed:
Removed corresponding-topic-id.

kaylahardie committed:
Merge branch 'mark-translations-rejected-for-topic-deletion' into mark-translation-rejected-when-exp-removed-from-story

kaylahardie committed:
Added asserts to test cases.

kaylahardie committed:
Split up tests in gae_models.py.

kaylahardie committed:
Merge branch 'mark-translations-rejected-for-topic-deletion' into mark-translation-rejected-when-exp-removed-from-story

kaylahardie committed:
Aligned bracket.

kaylahardie committed:
Merge remote-tracking branch 'upstream/develop' into mark-translations-rejected-for-topic-deletion

kaylahardie committed:
Merge branch 'mark-translations-rejected-for-topic-deletion' into mark-translation-rejected-when-exp-removed-from-story

kaylahardie committed:
Made testing messages more clear by adding a suggestion_id to the messages.

kaylahardie committed:
Merge remote-tracking branch 'upstream/develop' into mark-translations-rejected-for-topic-deletion

kaylahardie committed:
Fixed whitespace, added suggestion_id check.

kaylahardie committed:
Merge branch 'mark-translations-rejected-for-topic-deletion' into mark-translation-rejected-when-exp-removed-from-story

DubeySandeep commented on core/domain/suggestion_services.py:
```diff
@@ -151,6 +152,28 @@ def query_suggestions(query_fields_and_values):
+    Returns:
+        list(Suggestion). A list of translation suggestions that
+        correspond to the given exploration ids, up to a maximum of
+        feconf.DEFAULT_QUERY_LIMIT suggestions.
```
Does this limitation work for this method? if yes, then are we expecting this method to return all the corresponding suggestions which we will delete in the calling method?

kaylahardie replied:
I'm not sure if the limitation works for this method, I just saw the limitation applied to most methods so I thought I should keep it consistent and also apply it...maybe I should remove it? We do want to delete all the suggestions and not a subset. What do you think?

kaylahardie replied:
I changed it to make multiple fetches if necessary :)

DubeySandeep commented on core/domain/suggestion_services.py:
```diff
@@ -170,17 +193,37 @@ def _update_suggestion(suggestion):
 
-    suggestion_model = suggestion_models.GeneralSuggestionModel.get_by_id(
-        suggestion.suggestion_id)
+    _update_suggestions([suggestion])
```
Let's remove the _update_suggestion completely*

kaylahardie replied:
it's kind of nice to have it there as an api because a lot of the time we're just updating one suggestion. There's no code duplication because it's just calling _update_suggestions. What do you think? I took the same approach for my other methods.

DubeySandeep commented on core/domain/suggestion_services.py:
```diff
@@ -236,7 +279,11 @@ def accept_suggestion(suggestion, reviewer_id, commit_message, review_message):
-        raise Exception('The suggestion has already been accepted/rejected.')
+        raise Exception(
+            'The suggestion with id %s has already been accepted/rejected.' % (
+                suggestion.suggestion_id
```
suggestion.suggestion_id**))** (to reduce extra lines and lots of line with parenthesis, maybe? elsewhere as well)

seanlip replied:
Just to confirm -- either convention is fine. The rule that @kaylahardie is following is to have the closing paren on the same indentation level as the open paren, and that's OK too.

kaylahardie replied:
@DubeySandeep is this better? I moved one of the brackets up but kept the last one spaced because I think it improves readability.

DubeySandeep commented on core/domain/suggestion_services.py:
```diff
@@ -289,22 +340,48 @@ def reject_suggestion(suggestion, reviewer_id, review_message):
+
+    for suggestion in suggestions:
+        thread_id = suggestion.suggestion_id
+        feedback_services.create_message(
```
Will it make n db calls? If yes, can we do better?

kaylahardie replied:
Changed!

DubeySandeep commented on core/domain/suggestion_services.py:
```diff
@@ -318,10 +395,26 @@ def reject_question_suggestions_with_skill_target_id(skill_id):
+        suggestion_models.DELETED_SKILL_REJECT_MESSAGE)
+
+
+def reject_translation_suggestions_with_exp_target_ids(exp_ids):
```
Maybe autoreject_* or anything similar to highlight that this method will reject suggestions from a bot account*

kaylahardie replied:
Done :) Thanks for this suggestion.

DubeySandeep commented on core/domain/suggestion_services.py:
```diff
@@ -341,11 +434,18 @@ def resubmit_rejected_suggestion(suggestion, summary_message, author_id):
     if not suggestion.is_handled:
-        raise Exception('The suggestion is not yet handled.')
+        raise Exception(
+            'The suggestion with id %s is not yet handled.' % (
```
Thanks a lot for these changes!

kaylahardie replied:
No worries! :)

DubeySandeep commented on core/storage/suggestion/gae_models.py:
```diff
@@ -102,6 +102,14 @@
+# associated with an exploration that no longer corresponds to the story.
+# The story could be deleted or the exploration could have been
+# removed from the story.
+INVALID_STORY_REJECT_MESSAGE = (
```
INVALID_STORY_TRANSLATION_SUGGESTION_REJECT_MESSAGE (or something similar to highlight this will be only used for translation suggestions)

EDIT: The name I have to suggest seems long and I'm not happy with `INVALID_STORY_` prefix, so you may have to find another name.

kaylahardie replied:
Is INVALID_STORY_REJECT_TRANSLATION_SUGGESTIONS_MSG ok?

DubeySandeep reviewed (changes requested):
This looks much better now! I have a few more comments PTAL!

DubeySandeep commented on core/storage/suggestion/gae_models.py:
```diff
@@ -102,6 +102,14 @@
+# The story could be deleted or the exploration could have been
+# removed from the story.
+INVALID_STORY_REJECT_MESSAGE = (
+    'The associated story relationship is no longer valid.'
```
@seanlip, Do you have any suggestions on the message via a bot on rejecting suggestions? These translation suggestions are rejected as the corresponding exploration is not associated with any story/chapter!

seanlip replied:
Who's the audience? Is it the person who submitted the translation?

If so, then perhaps something like: "This text snippet has been removed from the story, and no longer needs translation. Sorry about that!" 

(Not sure ... what do you think? /cc @kaylahardie )

seanlip replied:
Also @DubeySandeep maybe you could ask Audrey? (Send her an email with the context.)

kaylahardie replied:
I think it could be sent to users because in create_messages it calls _add_message_to_email_buffer which sends emails if there are updates to the exploration.

kaylahardie replied:
If I'm right maybe I should change the DELETED_SKILL_REJECT_MESSAGE as well?

kaylahardie committed:
Made changes and gae_models_test still pass.

kaylahardie committed:
feedback_services, suggestion_services tests pass.

krishnarao22 commented:
Hi @kaylahardie, it seems like the e2e tests have flaked multiple times on this PR. Would you please report these flakes to Sandeep Dubey (with the logs of the failed tests)? This helps us fix common flakes. Thanks! :)

kaylahardie committed:
Addressed reviewers comments.

kaylahardie committed:
Merge remote-tracking branch 'upstream/develop' into mark-translations-rejected-for-topic-deletion

kaylahardie committed:
Fixed lint errors.

kaylahardie committed:
Condensed brackets.

kaylahardie committed:
resolved merge conflict.

kaylahardie committed:
Added auto to the method name.

kaylahardie reviewed (commented):
@DubeySandeep I made the changes, PTAL :)

kaylahardie commented:
@krishnarao22 I'll report the flakes, thanks :)

kaylahardie committed:
Fixed backend coverage.

kaylahardie committed:
Merge branch 'mark-translations-rejected-for-topic-deletion' into mark-translation-rejected-when-exp-removed-from-story

kaylahardie committed:
Addressed reviewers comments.

kaylahardie committed:
Adjusted tests for gae_models.

kaylahardie committed:
Fixed whitespace.

kaylahardie committed:
Fixed merge conflict.

kaylahardie committed:
Added return for create_messages.

kaylahardie committed:
Fixing an error.

kaylahardie committed:
Changed variable name.

kaylahardie commented:
@DubeySandeep I'm adding more changes to PR 10022 (which this branch branches off of) and I'm also going to fix the query limit. I'll add you to this PR again tomorrow after the changes are made, thanks!

kaylahardie committed:
Changed another variable name.

kaylahardie committed:
Addressed reviewers comments.

kaylahardie committed:
Merge remote-tracking branch 'upstream/develop' into mark-translations-rejected-for-topic-deletion

kaylahardie committed:
Changed docstring.

kaylahardie committed:
Changed docstring for one of the tests to be more concise.

kaylahardie committed:
Merge branch 'mark-translations-rejected-for-topic-deletion' into mark-translation-rejected-when-exp-removed-from-story

kaylahardie committed:
get_translation_suggestions_with_exp_ids can now fetch the datastore multiple times if the query results surpass the limit.

kaylahardie commented:
@DubeySandeep I made the changes PTAL :)

kaylahardie commented:
@seanlip @nithusha21 @aks681 this PR branches off from [PR 10022](https://github.com/oppia/oppia/pull/10022) (which you all reviewed). The files changed in this PR are:
- `core/storage/suggestion/gae_models.py`
- `core/storage/suggestion/gae_models_test.py`
- `core/domain/suggestion_services_test.py`
- `core/domain/story_services.py`

More specifically, the differences are:
- in `core/storage/suggestion/gae_models.py` I changed it so that `get_translation_suggestions_with_exp_ids` can fetch the datastore multiple times, with the `feconf.DEFAULT_QUERY_LIMIT` applied to each fetch. I added a test for this change in `core/storage/suggestion/gae_models_test.py` Both of these changes are in [commit 0aeabe036d7bfdb2355a1388bf136b717bcc2ee0](https://github.com/oppia/oppia/pull/10024/commits/0aeabe036d7bfdb2355a1388bf136b717bcc2ee0)
- I added `suggestion_services.auto_reject_translation_suggestions_for_exp_ids(exp_ids_removed_from_story)` in the `update_story` method in `story_services.py` to reject the suggestions when an exploration is removed from a story. This is also visible in the [commit 0aeabe036d7bfdb2355a1388bf136b717bcc2ee0](https://github.com/oppia/oppia/pull/10024/commits/0aeabe036d7bfdb2355a1388bf136b717bcc2ee0)
- I added tests to test that the suggestions were rejected when an exploration was removed from a story as well as when a story was deleted. These two tests are `test_delete_story_rejects_translation_suggestion` and `test_remove_exp_from_story_rejects_translation_suggestion` in `suggestion_services_test.py`. You can see the most updated version of those tests by viewing all of the commits (most of the code was written in [commit f153e2416f195e94e20d6a6badb59a298b3bcbd3](https://github.com/oppia/oppia/pull/10024/commits/f153e2416f195e94e20d6a6badb59a298b3bcbd3) but it was updated later)

Just thought I'd outline the differences to make it easier :) This PR is part of the next release so I'm trying to submit it as soon as possible.

kaylahardie committed:
Addressed final reviewers comments.

kaylahardie committed:
Merge branch 'mark-translations-rejected-for-topic-deletion' into mark-translation-rejected-when-exp-removed-from-story

kaylahardie committed:
Fetching results does not guarantee order.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -232,6 +241,34 @@ def query_suggestions(cls, query_fields_and_values):
+            suggestion_models.extend(results)
+        return suggestion_models
+
+
```
Drop one newline -- methods in a class should be separated by 1 newline, not 2.

kaylahardie replied:
Removed - thanks!

seanlip reviewed (commented):
Hi @kaylahardie -- thanks for signposting the files. I took a look and I don't really have any comments! It generally looks fine to me.

seanlip commented:
(Feel free to assign me again once #10022 is merged and I'll take a final pass and (probably) approve it.)

nithusha21 reviewed (approved):
I don't have any concerns... Thanks for fixing these issues!

kaylahardie committed:
Fixed merge conflicts.

kaylahardie committed:
Improved documentation.

kaylahardie committed:
Fixed another merge conflict.

kaylahardie committed:
One more minor docstring change.

kaylahardie reviewed (commented):
@seanlip I removed the newline and fixed the merge conflicts, PTAL :)

aks681 reviewed (approved):
Lgtm for one codeowner file.

seanlip reviewed (approved):
Thanks! LGTM, no issues from my end.

DubeySandeep reviewed (approved):
LGTM for the codeowner files!

DubeySandeep commented:
For unknown reasons, Github isn't showing the status of the Travis build though I see "Green tick" for the last commit. @seanlip Can we merge this PR?

EDIT:
@seanlip Ignore this comment I think Ben needs to review this PR! I'll ping/assign you to merge this PR (if needed).

kaylahardie commented:
@DubeySandeep thanks for your review & re your comment - I think it's because the green ticks next to the commit don't include the Travis build

seanlip commented:
Actually @BenHenning's review is optional. We just need the link to the Travis tests. Not sure why it's not registering, I will try looking for it.

seanlip commented:
Yup it all passes! Will merge. https://travis-ci.com/github/oppia/oppia/builds/178694249

Thanks @kaylahardie (and @DubeySandeep for flagging)!
